beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -2616,21 +2590,29 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
                 .map_err(BeaconChainError::from)?;
         }
 
-        let mut fork_choice = self.fork_choice.write();
+        let mut fork_choice = self.canonical_head.fork_choice_write_lock();
```

Comment:
It might be better to change this to
```rust
let fork_choice = self.canonical_head.fork_choice_read_lock();
```
and then wait until after the
```rust
if let Some(wss_checkpoint) = self.config.weak_subjectivity_checkpoint {
    ...
    let current_head_finalized_checkpoint =
        self.canonical_head.cached_head().finalized_checkpoint();
    ...
}
```
block below and insert:
```rust
drop(fork_choice);
let mut fork_choice = self.canonical_head.fork_choice_write_lock();
```

For 2 reasons:
1. Lock contention *slightly* reduced
2. Inside the `if` statement you call `self.canonical_head.cached_head()` which acquires the `cached_head` read lock. Now from what I can see, this isn't an issue right now. But if at some point in the future, a piece of code acquires the `cached_head` write lock, and then simultaneously tries to acquire *either* of the `fork_choice` read or write locks, we can have a deadlock.